app/com/linkedin/drelephant/tuning/hbt/PigHbtParameterRecommender.java

```diff
@@ -0,0 +1,406 @@
+            } else if (heuristicResultDetail.name.equals(SORT_FACTOR_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(SORT_FACTOR_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(PIG_MAX_SPLIT_SIZE_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(PIG_SPLIT_SIZE_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(SORT_SPILL_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(SORT_SPILL_HADOOP_CONF.getValue(),
+                  Double.parseDouble(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(MAPPER_HEAP_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(MAPPER_HEAP_HADOOP_CONF.getValue(), getHeapMemory(heuristicResultDetail.value));
+            } else if (heuristicResultDetail.name.equals(REDUCER_HEAP_HEURISTICS_CONF.getValue())) {
+              latestAppliedParams.put(REDUCER_HEAP_HADOOP_CONF.getValue(), getHeapMemory(heuristicResultDetail.value));
+            }
+          }
+        }
+      }
+    } else {
+      throw new IllegalArgumentException("No App Results found while TuneIn was suggesting new parameters");
+    }
+  }
+
+  /**
+   * This method is responsible for suggesting the parameters to fix failed heuristics
+   * in the next execution or if there is no failed heuristics then try to optimize Resource usage
+   * @return Map with suggested parameter's name as key and its value
+   */
+  public Map<String, Double> suggestParameters() {
+    loadLatestAppliedParameters();
+    List<String> failedHeuristicNameList = getFailedHeuristics();
+    suggestToFixFailedHeuristics(failedHeuristicNameList);
+    if (failedHeuristicNameList.size() == 0) {
+      logger.info("No failed heuristics, so will try to optimize memory parameters");
+      suggestMemoryParam(MRJobTaskType.MAP);
+      suggestMemoryParam(MRJobTaskType.REDUCE);
+    }
+    addRequiredPreviousSuggestedParameters();
+    modifyMapperMemory();
+    return jobSuggestedParameters;
+  }
+  /**
+   * Method to suggest parameters accordingly for each failed heuristic
+   * @param failedHeuristicNameList list with names of heuristics which are failed
+   */
+  private void suggestToFixFailedHeuristics(List<String> failedHeuristicNameList) {
+    for (String failedHeuristicName : failedHeuristicNameList) {
+      switch (failedHeuristicName) {
+        case MAPPER_MEMORY :
+          suggestMemoryParam(MRJobTaskType.MAP);
+          break;
+        case REDUCER_MEMORY :
+          suggestMemoryParam(MRJobTaskType.REDUCE);
+          break;
+        case MAPPER_SPILL :
+          suggestParametersForMemorySpill();
+          break;
+        case MAPPER_TIME:
+        case MAPPER_SPEED:
+          suggestSplitSize();
```

Comment:
SplitSize to be suggested if mapper time is failed . Mapper Speed heuristic generally requires change in code

Reply:
I think , we can have better exception handling , since in case there is error/exception (unknown) while suggesting mapper parameter , we should still be able to continue and do the processing for Reducer Memory . Ideally if the two heuristics are independent ,then failing for suggesting parameter for one should not cause any failure for suggesting parameter for another one

Reply:
`SplitSize to be suggested if mapper time is failed . Mapper Speed heuristic generally requires change in code` :: In the Mapper Speed heuristics severity is the min(mapper speed, number of tasks) so tuning #task would surely help in fixing the Mapper Speed heuristic.